### ISSUE TITLE
Bug: calc doesn't compile with musl libc on gentoo

### ISSUE THREAD

**Issue by akdjka** [lcn2/calc#61]
Calc bug report template version: 1.2 2022-02-24

**Describe the bug**
I tried to install calc on my gentoo. Build failed, I notice a long-standing bug:
https://bugs.gentoo.org/717694
So I downloaded the latest calc from master and tried to compile. It failed the same way.

**To Reproduce**
Download the latest git variant of calc (3e4391e), make.
There is an error:
```
➜  calc git:(25dcd8c) ✗ make clean debug 2>&1 > debug.log
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-gentoo-linux-musl/11.3.0/lto-wrapper
Target: x86_64-gentoo-linux-musl
Configured with: /var/tmp/portage/sys-devel/gcc-11.3.0/work/gcc-11.3.0/configure --host=x86_64-gentoo-linux-musl --build=x86_64-gentoo-linux-musl --prefix=/usr --bindir=/usr/x86_64-gentoo-linux-musl/gcc-bin/11.3.0 --includedir=/usr/lib/gcc/x86_64-gentoo-linux-musl/11.3.0/include --datadir=/usr/share/gcc-data/x86_64-gentoo-linux-musl/11.3.0 --mandir=/usr/share/gcc-data/x86_64-gentoo-linux-musl/11.3.0/man --infodir=/usr/share/gcc-data/x86_64-gentoo-linux-musl/11.3.0/info --with-gxx-include-dir=/usr/lib/gcc/x86_64-gentoo-linux-musl/11.3.0/include/g++-v11 --with-python-dir=/share/gcc-data/x86_64-gentoo-linux-musl/11.3.0/python --enable-languages=c,c++,fortran --enable-obsolete --enable-secureplt --disable-werror --with-system-zlib --disable-nls --disable-libunwind-exceptions --enable-checking=release --with-bugurl=https://bugs.gentoo.org/ --with-pkgversion='Gentoo 11.3.0 p7' --disable-esp --enable-libstdcxx-time --disable-libstdcxx-pch --enable-shared --enable-threads=posix --enable-__cxa_atexit --disable-multilib --with-multilib-list=m64 --disable-fixed-point --enable-targets=all --enable-libgomp --disable-libssp --disable-libada --disable-cet --disable-systemtap --disable-valgrind-annotations --disable-vtable-verify --disable-libvtv --without-zstd --enable-lto --without-isl --disable-libsanitizer --enable-default-pie --enable-default-ssp
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 11.3.0 (Gentoo 11.3.0 p7)
In file included from qmath.h:32,
                 from cmath.h:32,
                 from byteswap.c:28:
byteswap.c: In function 'swap_HALF_in_ZVALUE':
zmath.h:581:36: warning: '*dest.v' may be used uninitialized [-Wmaybe-uninitialized]
  581 | #define zcopyval(z1,z2) memcpy((z2).v, (z1).v, (z1).len * sizeof(HALF))
      |                                ~~~~^~
byteswap.c:543:36: note: in expansion of macro 'zcopyval'
  543 |                 dest->v = (HALF *) zcopyval(*src, *dest);
      |                                    ^~~~~~~~
In file included from file.c:44:
file.c: In function 'filepos2z':
fposval.h:16:54: error: 'FPOS_POS_LEN' undeclared (first use in this function); did you mean 'FILEPOS_LEN'?
   16 |         memcpy((void *)(dest), (void *)(src), sizeof(FPOS_POS_LEN))
      |                                                      ^~~~~~~~~~~~
file.c:1391:9: note: in expansion of macro 'SWAP_HALF_IN_FILEPOS'
 1391 |         SWAP_HALF_IN_FILEPOS(ret.v, &pos);
      |         ^~~~~~~~~~~~~~~~~~~~
fposval.h:16:54: note: each undeclared identifier is reported only once for each function it appears in
   16 |         memcpy((void *)(dest), (void *)(src), sizeof(FPOS_POS_LEN))
      |                                                      ^~~~~~~~~~~~
file.c:1391:9: note: in expansion of macro 'SWAP_HALF_IN_FILEPOS'
 1391 |         SWAP_HALF_IN_FILEPOS(ret.v, &pos);
      |         ^~~~~~~~~~~~~~~~~~~~
file.c: In function 'z2filepos':
file.c:1462:67: error: expected ')' before ';' token
 1462 |                 memcpy(&tmp, zpos.v, MIN(sizeof(tmp), FILEPOS_LEN);
      |                       ~                                           ^
file.c:1462:68: error: expected ';' before '}' token
 1462 |                 memcpy(&tmp, zpos.v, MIN(sizeof(tmp), FILEPOS_LEN);
      |                                                                    ^
      |                                                                    ;
 1463 |         } else {
      |         ~
file.c:60:24: warning: comparison of integer expressions of different signedness: 'long unsigned int' and 'int' [-Wsign-compare]
   60 | #define MIN(a,b) (((a) <= (b)) ? (a) : (b))
      |                        ^~
file.c:1466:38: note: in expansion of macro 'MIN'
 1466 |                 memcpy(&tmp, zpos.v, MIN(sizeof(tmp),
      |                                      ^~~
file.c:60:40: warning: operand of '?:' changes signedness from 'int' to 'long unsigned int' due to unsignedness of other operand [-Wsign-compare]
   60 | #define MIN(a,b) (((a) <= (b)) ? (a) : (b))
      |                                        ^~~
file.c:1466:38: note: in expansion of macro 'MIN'
 1466 |                 memcpy(&tmp, zpos.v, MIN(sizeof(tmp),
      |                                      ^~~
In file included from file.c:44:
fposval.h:16:54: error: 'FPOS_POS_LEN' undeclared (first use in this function); did you mean 'FILEPOS_LEN'?
   16 |         memcpy((void *)(dest), (void *)(src), sizeof(FPOS_POS_LEN))
      |                                                      ^~~~~~~~~~~~
file.c:1470:9: note: in expansion of macro 'SWAP_HALF_IN_FILEPOS'
 1470 |         SWAP_HALF_IN_FILEPOS(&ret, &tmp);
      |         ^~~~~~~~~~~~~~~~~~~~
make[1]: *** [Makefile:2688: file.o] Error 1
make: *** [Makefile:4964: debug] Error 2
```

**Attach make debug output**

[debug.log](https://github.com/lcn2/calc/files/10099447/debug.log)

**Execution environment (please complete the following information):**
 - OS:
        Gentoo Linux with musl and clang

  - Calc Version
        [v2.14.1.1](https://github.com/lcn2/calc/releases/tag/v2.14.1.1)

**Comment by lcn2** [comment]
hummm ....

**Comment by lcn2** [comment]
Please try commit 827988c55323f1067566d6304c157601a7d7f445 -- just checked in.

We appreciate you, @akdjka, submitting your bug report.  If you find a problem, please submit an updated bug report.

Thank you!

**Comment by akdjka** [comment]
I don't have this file.

**Comment by lcn2** [comment]
Thanks you ver much @akdjka for your most recent debug.log.

With commit 340c1990ce1774c964839fe3495af8ca4bfb617c we have moved towards better support for musl libc under gentoo.

We plan to address a technical issue with file positioning (this is related to who `file.c` filed to compile for you) in calc.  Calc treats file position as an integer numeric value, which is technically a _no-no_.  Nevertheless, for historic reasons calc treats file positions using integer numeric values, and as such calc has to make use of the implementation defined `fpos_t` type in non-standard ways.

The implementation of musl libc has a different implementation of the `fpos_t` type.  This combined with calc treating file positions using integer numeric values means we have to accommodate how musl libc uses `fpos_t`.  We plan to address this very soon.

Please try the top of the master branch under musl libc on gentoo.  If you find another problem, please submit an updated bug report with a new `debug.log` file.

Thanks for your help!

**Comment by akdjka** [comment]
Calc compiles and works on my system now. :)
Thank you!

**Comment by lcn2** [comment]
> Calc compiles and works on my system now. :)
> Thank you!

Yea!  Thanks @akdjka !

**QUESTION**: Does it pass `make chk`? 

If it does pass for you, we will roll out the code now as _calc version 2.14.1.2_.

We ask that this bug be left open for a bit longer.  We are working on better fix that will
make musl libc happier AND will better handle other libc implementations of `fpos_t` as well.
This will become the basis of _calc version 2.14.2.0_.

So you don't mind testing that version when it comes out, we would appreciate it.

**Comment by akdjka** [comment]
Unfortunately, it fails.
```
➜  calc make chk
CALCPATH=./cal LD_LIBRARY_PATH=. DYLD_LIBRARY_PATH=. CALCHELP=./help CALCCUSTOMHELP=./custom ./calc -d -q read regress 2>&1 | awk -f check.awk
4600: Beginning test_fileop
4601: ttest(0,0): passed
4602: stest(): failed
**** search(f, "and") != 10 failed
4603: ttest(1,1): failed
**** Failure 1 for file size
4604: stest(): failed
**** search(f, "and") != 10 failed
4605: ttest(2,2): failed
**** Failure 1 for file size
4606: stest(): failed
**** search(f, "and") != 10 failed
4607: ttest(3,3): failed
**** Failure 1 for file size
4608: stest(): failed
**** search(f, "and") != 10 failed
4609: ttest(4,4): failed
**** Failure 1 for file size
4610: stest(): failed
**** search(f, "and") != 10 failed
4611: ttest(5,5): failed
**** Failure 1 for file size
4612: stest(): failed
**** search(f, "and") != 10 failed
4613: ttest(6,6): failed
**** Failure 1 for file size
4614: stest(): failed
**** search(f, "and") != 10 failed
4615: ttest(7,7): failed
**** Failure 1 for file size
4616: stest(): failed
**** search(f, "and") != 10 failed
4617: ttest(8,8): failed
**** Failure 1 for file size
4618: stest(): failed
**** fopen("junk4600", "wb") failed
4619: ttest(9,9): failed
**** Failure 1 for file size
4620: stest(): failed
**** fopen("junk4600", "wb") failed
4621: Ending test_fileop

4700: Beginning test_charset
**** errcount:182  > ecnt:159
4701: "\a" == char(7)
4702: "\v" == char(11)
4703: "\e" == char(27)
5000: Beginning test_filesearch
5001: x = rm("-f", "junk5000")
5002: f = fopen("junk5000", "w")
**** Unable to open "junk5000" for writing

5100: Beginning test_newdecl
5101: test5100(1)
**** errcount:183  > ecnt:182
5102: a5100 == 0
5103: b5100 == 2
5104: test5100(1)
5927: test unused
5928: test unused
5929: test unused
**** errcount:190  > ecnt:187
5930: isassoc(loc) == 0
5931: isassoc(a) == 1
5932: isassoc(ofd) == 0
6079: test unused
6080: iserror(loc) == 0
6081: iserror(a) == 0
**** Non-true result (0): 6082: iserror(ofd) == 0
**** Non-true result (0): 6083: iserror(cfd) == 0
6084: iserror(blk) == 0
6085: iserror(nblk) == 0
6086: iserror(cfg) == 0
6139: test unused
6140: isfile(loc) == 0
6141: isfile(a) == 0
**** Non-true result (0): 6142: isfile(ofd) == 1
**** Non-true result (0): 6143: isfile(cfd) == 1
6144: isfile(blk) == 0
6145: isfile(nblk) == 0
6146: isfile(cfg) == 0

6700: Beginning test_blk
6701: A = blk(20);
**** errcount:191  > ecnt:190
6702: size(A) == 20
6703: sizeof(A) == 256
6704: B = A;
6822: C == A
6823: fs = fopen("junk6800", "w+");
6824: blkcpy(fs, A);
**** errcount:193  > ecnt:191
**** Non-true result (0): 6825: size(f) == 5
6826: blkcpy(B = blk(), fs);
**** errcount:194  > ecnt:193
**** Non-true result (0): 6827: B == A
6828: blkcpy(fs, A, ,100);
**** errcount:195  > ecnt:194
**** Non-true result (0): 6829: size(f) == 105
6830: blkcpy(C = blk(), fs,2,,100)
**** errcount:196  > ecnt:195
**** Non-true result (0): 6831: C == (blk() = {1,2}
6832: A = blk();
6833: blkcpy(A, "blk6800");
6834: size(A) == 9
6900: Beginning test_name
6901: x = rm("-f", "junk6900")
6902: f = fopen("junk6900", "w")
**** errcount:199  > ecnt:196
**** Non-true result (0): 6903: name(f) == "junk6900"
6904: fclose(f)
**** errcount:200  > ecnt:199
6905: name(f) == null()
6906: A = blk("blk6900")
6907: name(A) == "blk6900"
7002: B = blk();
7003: copy("abc yz", A);
7004: copy("defg", B);
**** errcount:201  > ecnt:200
7005: strprintf("%s", A) == "abc yz"
7006: strprintf("%s", A[2]) == "c yz"
7007: strprintf("%s", A[7]) == ""

9995: freeredc()
9996: freestatics()
**** 21 error(s) found  \/++\/
9998: freeglobals()
9999: Ending regression tests
make: *** [Makefile:4684: chk] Error 1
```

**Comment by lcn2** [comment]
> Unfortunately, it fails.
> 
> ```
> 
> ➜  calc make chk
> 
> CALCPATH=./cal LD_LIBRARY_PATH=. DYLD_LIBRARY_PATH=. CALCHELP=./help CALCCUSTOMHELP=./custom ./calc -d -q read regress 2>&1 | awk -f check.awk
> 
> 4600: Beginning test_fileop
> 
> 4601: ttest(0,0): passed
> 
> 4602: stest(): failed
> 
> **** search(f, "and") != 10 failed
> 
> 4603: ttest(1,1): failed
> 
> **** Failure 1 for file size
> 
> 4604: stest(): failed
> 
> **** search(f, "and") != 10 failed
> 
> 4605: ttest(2,2): failed
> 
> **** Failure 1 for file size
> 
> 4606: stest(): failed
> 
> **** search(f, "and") != 10 failed
> 
> 4607: ttest(3,3): failed
> 
> **** Failure 1 for file size
> 
> 4608: stest(): failed
> 
> **** search(f, "and") != 10 failed
> 
> 4609: ttest(4,4): failed
> 
> **** Failure 1 for file size
> 
> 4610: stest(): failed
> 
> **** search(f, "and") != 10 failed
> 
> 4611: ttest(5,5): failed
> 
> **** Failure 1 for file size
> 
> 4612: stest(): failed
> 
> **** search(f, "and") != 10 failed
> 
> 4613: ttest(6,6): failed
> 
> **** Failure 1 for file size
> 
> 4614: stest(): failed
> 
> **** search(f, "and") != 10 failed
> 
> 4615: ttest(7,7): failed
> 
> **** Failure 1 for file size
> 
> 4616: stest(): failed
> 
> **** search(f, "and") != 10 failed
> 
> 4617: ttest(8,8): failed
> 
> **** Failure 1 for file size
> 
> 4618: stest(): failed
> 
> **** fopen("junk4600", "wb") failed
> 
> 4619: ttest(9,9): failed
> 
> **** Failure 1 for file size
> 
> 4620: stest(): failed
> 
> **** fopen("junk4600", "wb") failed
> 
> 4621: Ending test_fileop
> 
> 
> 
> 4700: Beginning test_charset
> 
> **** errcount:182  > ecnt:159
> 
> 4701: "\a" == char(7)
> 
> 4702: "\v" == char(11)
> 
> 4703: "\e" == char(27)
> 
> 5000: Beginning test_filesearch
> 
> 5001: x = rm("-f", "junk5000")
> 
> 5002: f = fopen("junk5000", "w")
> 
> **** Unable to open "junk5000" for writing
> 
> 
> 
> 5100: Beginning test_newdecl
> 
> 5101: test5100(1)
> 
> **** errcount:183  > ecnt:182
> 
> 5102: a5100 == 0
> 
> 5103: b5100 == 2
> 
> 5104: test5100(1)
> 
> 5927: test unused
> 
> 5928: test unused
> 
> 5929: test unused
> 
> **** errcount:190  > ecnt:187
> 
> 5930: isassoc(loc) == 0
> 
> 5931: isassoc(a) == 1
> 
> 5932: isassoc(ofd) == 0
> 
> 6079: test unused
> 
> 6080: iserror(loc) == 0
> 
> 6081: iserror(a) == 0
> 
> **** Non-true result (0): 6082: iserror(ofd) == 0
> 
> **** Non-true result (0): 6083: iserror(cfd) == 0
> 
> 6084: iserror(blk) == 0
> 
> 6085: iserror(nblk) == 0
> 
> 6086: iserror(cfg) == 0
> 
> 6139: test unused
> 
> 6140: isfile(loc) == 0
> 
> 6141: isfile(a) == 0
> 
> **** Non-true result (0): 6142: isfile(ofd) == 1
> 
> **** Non-true result (0): 6143: isfile(cfd) == 1
> 
> 6144: isfile(blk) == 0
> 
> 6145: isfile(nblk) == 0
> 
> 6146: isfile(cfg) == 0
> 
> 
> 
> 6700: Beginning test_blk
> 
> 6701: A = blk(20);
> 
> **** errcount:191  > ecnt:190
> 
> 6702: size(A) == 20
> 
> 6703: sizeof(A) == 256
> 
> 6704: B = A;
> 
> 6822: C == A
> 
> 6823: fs = fopen("junk6800", "w+");
> 
> 6824: blkcpy(fs, A);
> 
> **** errcount:193  > ecnt:191
> 
> **** Non-true result (0): 6825: size(f) == 5
> 
> 6826: blkcpy(B = blk(), fs);
> 
> **** errcount:194  > ecnt:193
> 
> **** Non-true result (0): 6827: B == A
> 
> 6828: blkcpy(fs, A, ,100);
> 
> **** errcount:195  > ecnt:194
> 
> **** Non-true result (0): 6829: size(f) == 105
> 
> 6830: blkcpy(C = blk(), fs,2,,100)
> 
> **** errcount:196  > ecnt:195
> 
> **** Non-true result (0): 6831: C == (blk() = {1,2}
> 
> 6832: A = blk();
> 
> 6833: blkcpy(A, "blk6800");
> 
> 6834: size(A) == 9
> 
> 6900: Beginning test_name
> 
> 6901: x = rm("-f", "junk6900")
> 
> 6902: f = fopen("junk6900", "w")
> 
> **** errcount:199  > ecnt:196
> 
> **** Non-true result (0): 6903: name(f) == "junk6900"
> 
> 6904: fclose(f)
> 
> **** errcount:200  > ecnt:199
> 
> 6905: name(f) == null()
> 
> 6906: A = blk("blk6900")
> 
> 6907: name(A) == "blk6900"
> 
> 7002: B = blk();
> 
> 7003: copy("abc yz", A);
> 
> 7004: copy("defg", B);
> 
> **** errcount:201  > ecnt:200
> 
> 7005: strprintf("%s", A) == "abc yz"
> 
> 7006: strprintf("%s", A[2]) == "c yz"
> 
> 7007: strprintf("%s", A[7]) == ""
> 
> 
> 
> 9995: freeredc()
> 
> 9996: freestatics()
> 
> **** 21 error(s) found  \/++\/
> 
> 9998: freeglobals()
> 
> 9999: Ending regression tests
> 
> make: *** [Makefile:4684: chk] Error 1
> 
> ```

This is probably due to the previously mentioned `fpos_t` issues that we plan to address in the next version of calc.

**Comment by lcn2** [comment]
Just to let you know, @akdjka  we are experiencing Internet connectivity problems that may last through Tuesday, for some 24 hour value of Tuesday :), and thus it may be several days before we can address your recent information. 

Thank you for updating us, and we hope to, when Internet services restored, get back to your issue soon. Sorry (tm Canada 🇨🇦) for the inconvenience and thanks for your patience.

**Comment by lcn2** [comment]
Just an update, @akdjka  

We will make this a priority .. although right now we have a regional internet service outage .. this reply is being send thru a satellite phone ($$$-ive).  Nevertheless we wanted to let you know that we will address this issue, work offline on the matter, and when our server is restored (in a few days is the ETA) we will get back to you.

**Comment by akdjka** [comment]
Thank you. Still, calc actually seems to be working for me. Since there are probably few people being blocked by it, I don't think it deserves being a high priority.

**Comment by lcn2** [comment]
It turns out that a new issue #63 is related to this issue (see [comment 1336220653](https://github.com/lcn2/calc/issues/63#issuecomment-1336220653)).

We plan to address this musl libc / fpos_t / filepos2z()&z2filepos()` next, as part of our calc version 2.14.2.0 release and as part of fixing this bug.  As such we wish to keep this bug report an open priority.

**Comment by lcn2** [comment]
This issue was completed back in 2022 Dec.
